File: ruby/bowling/_iterations/1/bowling.rb
class Game
  class BowlingError < StandardError; end

  def initialize
    @frames = [Frame.new]
  end

  def done?
    @frames.size == 10 && @frames.last.done?
  end

  def roll(pins)
    raise BowlingError, "Game is done" if done?
    raise BowlingError, "Roll must knock down 0 or more pins" if pins.negative?

    @frames.last.roll(pins)
    if @frames.last.done? && @frames.size < 10
      @frames << Frame.new(last: @frames.size == 9)
    end
  end

  def score
    raise BowlingError, "Cannot calculate score until game is done" unless done?

    @frames.map.with_index do |frame, index|
      frame.score(@frames.drop(index + 1).map(&:rolls).reduce(&:+))
    end.sum
  end

  private

  class Frame
    attr_reader :rolls, :last

    def initialize(last: false)
      @rolls = []
      @max_rolls = 2
      @last = last
    end

    def done?
      rolls.size == @max_rolls || (!last && rolls.first == 10)
    end

    def roll(pins)
      raise BowlingError, "Single roll cannot knock down more than 10 pins" if pins > 10
      raise BowlingError, "Frame cannot knock down more than 10 pins" if !last && (rolls.sum + pins) > 10
      raise BowlingError, "Last frame cannot knock down more than 10 pins if spare" if last && rolls.size == 1 && rolls.first < 10 && rolls.first + pins > 10
      raise BowlingError, "Last frame cannot knock down more than 10 pins if bonus is spare" if last && rolls.size == 2 && rolls.first == 10 && rolls[1] < 10 && rolls[1] + pins > 10

      rolls << pins
      if last && @max_rolls == 2 && rolls.sum == 10
        @max_rolls = 3
      end
    end

    def score(next_rolls)
      if !last && rolls.first == 10
        10 + next_rolls[0, 2].sum
      elsif !last && rolls.sum == 10
        10 + next_rolls[0, 1].sum
      else
        rolls.sum
      end
    end
  end
end

File: javascript/bowling/_iterations/1/bowling.js
export class Bowling {
    constructor() {
        this.frames = [
            new Frame(),
        ];
        this.done = false;
    }

    roll(score) {
        if (this.done) {
            throw new Error('Cannot roll after game is over');
        } else if (score < 0) {
            throw new Error('Negative roll is invalid');
        }

        const frame = this.frames[this.frames.length - 1];
        frame.roll(score);
        if (frame.done) {
            if (this.frames.length < 9) {
                this.frames.push(new Frame());
            } else if (this.frames.length === 9) {
                this.frames.push(new LastFrame());
            } else {
                this.done = true;
            }
        }
    }

    score() {
        if (!this.done) {
            throw new Error('Score cannot be taken until the end of the game')
        }

        return this.frames.map((frame, idx, frames) => {
            const nextTwoThrows = [];
            if (idx !== 10 && (idx  + 1) < frames.length) {
                nextTwoThrows.push(frames[idx + 1].rolls[0]);
                if (frames[idx + 1].rolls.length > 1) {
                    nextTwoThrows.push(frames[idx + 1].rolls[1]);
                } else if ((idx + 2) < frames.length) {
                    nextTwoThrows.push(frames[idx + 2].rolls[0]);
                }
            }
            return frame.score(nextTwoThrows);
        }).reduce((acc, i) => acc + i, 0);
    }
}

class Frame {
    constructor() {
        this.rolls = [];
    }

    get done() {
        return this.rolls.length === 2 || (this.rolls.length === 1 && this.rolls[0] === 10);
    }

    roll(score) {
        if (score + (this.rolls.length > 0 ? this.rolls[0] : 0) > 10) {
            throw new Error('Pin count exceeds pins on the lane');
        }

        this.rolls.push(score);
    }

    score(nextTwoThrows) {
        if (this.rolls[0] === 10) {
            return 10 + nextTwoThrows[0] + nextTwoThrows[1];
        } else if ((this.rolls[0] + this.rolls[1]) === 10) {
            return 10 + nextTwoThrows[0];
        } else {
            return this.rolls[0] + this.rolls[1];
        }
    }
}

class LastFrame {
    constructor() {
        this.rolls = [];
        this.remain = 10;
    }

    get done() {
        return this.rolls.length === 3 || (this.rolls.length === 2 && (this.rolls[0] + this.rolls[1]) < 10);
    }

    roll(score) {
        if (score > this.remain) {
            throw new Error('Pin count exceeds pins on the lane');
        }

        this.rolls.push(score);
        this.remain -= score;
        if (this.remain === 0) {
            this.remain = 10;
        }
    }

    score() {
        return this.rolls.reduce((acc, i) => acc + i, 0);
    }
}
